packages/common-all/src/engine/EngineV3Base.ts

```diff
@@ -382,4 +392,88 @@ export abstract class EngineV3Base implements ReducedDEngine {
       })
     );
   }
+
+  /**
+   * Create backlink from given link that references another note (denoted by presence of link.to field)
+   * and add that backlink to referenced note's links
+   *
+   * @param link Link potentionally referencing another note
+   */
+  protected async addBacklink(link: DLink): Promise<NoteChangeEntry[]> {
+    const maybeBacklink = BacklinkUtils.createFromDLink(link);
+    if (maybeBacklink) {
+      const maybeVault = link.to?.vaultName
+        ? VaultUtils.getVaultByName({
+            vname: link.to?.vaultName,
+            vaults: this.vaults,
+          })
+        : undefined;
+      const notes = await this.noteStore.findMetaData({
+        fname: link.to!.fname!,
+        vault: maybeVault,
+      });
+      if (notes.data) {
+        return Promise.all(
+          notes.data.map(async (note) => {
+            const prevNote = _.merge(_.cloneDeep(note), {
+              body: "",
+            });
+            BacklinkUtils.addBacklink({ note, backlink: maybeBacklink });
+            // Temp solution to get around current restrictions where NoteChangeEntry needs a NoteProp
+            const updatedNote = _.merge(note, {
+              body: "",
+            });
+            return {
+              prevNote,
+              note: updatedNote,
+              status: "update",
+            };
+          })
+        );
+      }
+    }
+    return [];
+  }
+
+  /**
+   * Remove backlink associated with given link that references another note (denoted by presence of link.to field)
+   * from that referenced note
+   *
+   * @param link Link potentionally referencing another note
+   */
+  protected async removeBacklink(link: DLink): Promise<NoteChangeEntry[]> {
+    const maybeBacklink = BacklinkUtils.createFromDLink(link);
+    if (maybeBacklink) {
+      const maybeVault = link.to?.vaultName
+        ? VaultUtils.getVaultByName({
+            vname: link.to?.vaultName,
+            vaults: this.vaults,
+          })
+        : undefined;
+      const notes = await this.noteStore.findMetaData({
+        fname: link.to!.fname!,
```

Comment:
instead of `!` - can we do that same thing again as in the other review? If those aren't defined, do an early return of [] to save some backend calls.

Reply:
ah good reminder